### ISSUE TITLE
BUG: lax.conv_.. outputs are slightly different than Torch counterparts

### ISSUE THREAD

**Issue by paganpasta** [google/jax#11790]
### Description

Hi, 

I am trying to compare results of a convolution2d operation in PyTorch and Jax. The results seem to be slightly different and was wondering if this mismatch is normal? Or perhaps, there is a better way to do such a comparison.

The last assert fails with the default tolerance. `dtypes` are `float32` for both the frameworks.

```python
import jax
import jax.numpy as jnp
import jax.lax as lax

import torch.nn as tnn
import torch

import numpy as np


def test_conv2d():
    random_image = jax.random.uniform(key=jax.random.PRNGKey(0), shape=(1, 3, 224, 224))
    t_cv = tnn.Conv2d(3, 3, kernel_size=3)
    j_w, j_b = jnp.asarray(t_cv.weight.detach().numpy()), jnp.asarray(t_cv.bias.detach().numpy().reshape((3, 1, 1)))

    assert jnp.equal(j_w, t_cv.weight.detach().numpy()).all()
    assert jnp.equal(j_b, t_cv.bias.detach().numpy().reshape(3, 1, 1)).all()

    j_out = lax.conv_general_dilated(
        lhs=random_image,
        rhs=j_w,
        window_strides=(1, 1),
        padding=((0, 0), (0, 0)),
        rhs_dilation=(1, 1),
        feature_group_count=1,
        precision=jax.lax.Precision.HIGHEST
    )
    j_out = j_out + j_b
    with torch.no_grad():
        t_out = jnp.asarray(t_cv(torch.tensor(np.asarray(random_image))).numpy())

    assert t_out.shape == j_out.shape
    assert jnp.isclose(t_out, j_out).all()


```

### What jax/jaxlib version are you using?

0.3.15

### Which accelerator(s) are you using?

- [X] CPU
- [ ] GPU
- [ ] TPU

**Comment by hawkinsp** [comment]
I can't reproduce this. Can you share what output you got? (And please verify this is on CPU?)

A small amount of variation between systems is normal and expected due to things like orders of operations.

**Comment by paganpasta** [comment]
Hi @hawkinsp, thanks for looking into this.  

Yes, it is on the CPU. 
The output ndarrays are attached. To be loaded with `jnp.load(f)`
[outputs.zip](https://github.com/google/jax/files/9281952/outputs.zip)

So when you say it is not reproduced does it mean that the assertion passes?

**Comment by hawkinsp** [comment]
Correct: the assertion passes for me on every machine I have access to.

By share the outputs I just meant: what does the script print? How large is the error?

**Comment by paganpasta** [comment]
Oh I see, If I lower the tolerance 'atol=1e-7` the check passes. 
The issue is that I am porting weights from PyTorch and during debugging I saw that differences start off small and then eventually diverge as the depth increases.

**Comment by hawkinsp** [comment]
Yes, I'd suggest that simply means that your test tolerances are too tight. Feel free to reopen if you see large numerical differences!

**Comment by patrick-kidger** [comment]
@paganpasta do I understand correctly that this issue is preventing porting e.g. an Inception net from PyTorch to JAX? Or something else? Can you share an example for which the divergence between models is nontrivial?

**Comment by paganpasta** [comment]
@patrick-kidger At the moment, the results of a ported VGG-11 varied greatly from torchvision implementation (~30% top-1 accuracy). 
As I am trying to debug layer by layer, it seems that `1e-7` to `1e-4` difference is expected (?). The bulk of the difference arises after a specific pooling module (`adaptive pooling`). However, this behaviour is not replicated by other ported networks which perform under +- 0.5% variance. 

I'll go further into this to try and isolate the issue. 

Thanks for your help!

**Comment by patrick-kidger** [comment]
Right. The differences between convolutions seem unfortunate but perhaps not too big of a deal.

The differences under adaptive pooling are definitely expected, as Equinox and PyTorch consciously choose to do these differently.
